// @keystone-6/core/dist/createApolloServer-23a6674d.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import { ApolloServer } from 'apollo-server-micro';
import { ApolloServer as ApolloServer$1 } from 'apollo-server-express';
import { ApolloServerPluginLandingPageGraphQLPlayground, ApolloServerPluginLandingPageDisabled } from 'apollo-server-core';
import { createSessionContext } from '../session/dist/keystone-6-core-session.esm.js';

const createApolloServerMicro = _ref => {
  let {
    graphQLSchema,
    createContext,
    sessionStrategy,
    graphqlConfig,
    connectionPromise
  } = _ref;

  const context = async _ref2 => {
    let {
      req,
      res
    } = _ref2;
    await connectionPromise;
    return createContext({
      sessionContext: sessionStrategy ? await createSessionContext(sessionStrategy, req, res, createContext) : undefined,
      req
    });
  };

  const serverConfig = _createApolloServerConfig({
    graphQLSchema,
    graphqlConfig
  });

  return new ApolloServer(_objectSpread(_objectSpread({}, serverConfig), {}, {
    context
  }));
};
const createApolloServerExpress = _ref3 => {
  let {
    graphQLSchema,
    createContext,
    sessionStrategy,
    graphqlConfig
  } = _ref3;

  const context = async _ref4 => {
    let {
      req,
      res
    } = _ref4;
    return createContext({
      sessionContext: sessionStrategy ? await createSessionContext(sessionStrategy, req, res, createContext) : undefined,
      req
    });
  };

  const serverConfig = _createApolloServerConfig({
    graphQLSchema,
    graphqlConfig
  });

  return new ApolloServer$1(_objectSpread(_objectSpread({}, serverConfig), {}, {
    context
  }));
};

const _createApolloServerConfig = _ref5 => {
  var _graphqlConfig$playgr;

  let {
    graphQLSchema,
    graphqlConfig
  } = _ref5;
  const apolloConfig = graphqlConfig === null || graphqlConfig === void 0 ? void 0 : graphqlConfig.apolloConfig;
  const playgroundOption = (_graphqlConfig$playgr = graphqlConfig === null || graphqlConfig === void 0 ? void 0 : graphqlConfig.playground) !== null && _graphqlConfig$playgr !== void 0 ? _graphqlConfig$playgr : process.env.NODE_ENV !== 'production';
  return _objectSpread(_objectSpread({
    schema: graphQLSchema,
    debug: graphqlConfig === null || graphqlConfig === void 0 ? void 0 : graphqlConfig.debug
  }, apolloConfig), {}, {
    plugins: playgroundOption === 'apollo' ? apolloConfig === null || apolloConfig === void 0 ? void 0 : apolloConfig.plugins : [playgroundOption ? ApolloServerPluginLandingPageGraphQLPlayground({
      settings: {
        'request.credentials': 'same-origin'
      }
    }) : ApolloServerPluginLandingPageDisabled(), ...((apolloConfig === null || apolloConfig === void 0 ? void 0 : apolloConfig.plugins) || [])],
    formatError: formatError(graphqlConfig)
  });
};

const formatError = graphqlConfig => {
  return err => {
    var _graphqlConfig$apollo;

    let debug = graphqlConfig === null || graphqlConfig === void 0 ? void 0 : graphqlConfig.debug;

    if (debug === undefined) {
      debug = process.env.NODE_ENV !== 'production';
    }

    if (!debug && err.extensions) {
      // Strip out any `debug` extensions
      delete err.extensions.debug;
      delete err.extensions.exception;
    }

    if (graphqlConfig !== null && graphqlConfig !== void 0 && (_graphqlConfig$apollo = graphqlConfig.apolloConfig) !== null && _graphqlConfig$apollo !== void 0 && _graphqlConfig$apollo.formatError) {
      return graphqlConfig.apolloConfig.formatError(err);
    } else {
      return err;
    }
  };
};

export { createApolloServerExpress as a, createApolloServerMicro as c };
