// @keystone-6/core/dist/json-field-type-polyfill-for-sqlite-2020a626.esm.js
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import { p as postgresql, s as sqlite, m as mysql, f as fieldType } from './mysql-673b47d5.esm.js';
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import './graphql-ts-schema-57db0426.esm.js';
import { fields, field } from '@graphql-ts/schema/api-with-context';

const _excluded = ["mode"];
const objectEntriesButAssumeNoExtraProperties = Object.entries;

function internalResolveFilter(entries, mode) {
  const entry = entries.shift();
  if (entry === undefined) return {};
  const [key, val] = entry;

  if (val == null) {
    return {
      AND: [{
        [key]: val
      }, internalResolveFilter(entries, mode)]
    };
  }

  switch (key) {
    case 'equals':
    case 'lt':
    case 'lte':
    case 'gt':
    case 'gte':
    case 'in':
    case 'contains':
    case 'startsWith':
    case 'endsWith':
      {
        return {
          AND: [{
            [key]: val,
            mode
          }, {
            not: null
          }, internalResolveFilter(entries, mode)]
        };
      }

    case 'notIn':
      {
        return {
          AND: [{
            NOT: [internalResolveFilter(objectEntriesButAssumeNoExtraProperties({
              in: val
            }), mode)]
          }, internalResolveFilter(entries, mode)]
        };
      }

    case 'not':
      {
        return {
          AND: [{
            NOT: [internalResolveFilter(objectEntriesButAssumeNoExtraProperties(val), mode)]
          }, internalResolveFilter(entries, mode)]
        };
      }
  }
}

function resolveCommon(val) {
  if (val === null) return null;
  return internalResolveFilter(objectEntriesButAssumeNoExtraProperties(val), undefined);
}
function resolveString(val) {
  if (val === null) return null;

  let {
    mode
  } = val,
      rest = _objectWithoutProperties(val, _excluded);

  return internalResolveFilter(objectEntriesButAssumeNoExtraProperties(rest), mode);
}

var filters = /*#__PURE__*/Object.freeze({
  __proto__: null,
  postgresql: postgresql,
  sqlite: sqlite,
  mysql: mysql,
  resolveCommon: resolveCommon,
  resolveString: resolveString
});

function mapOutputFieldToSQLite(field$1) {
  const innerResolver = field$1.resolve || (_ref => {
    let {
      value
    } = _ref;
    return value;
  });

  return fields()({
    value: field({
      type: field$1.type,
      args: field$1.args,
      deprecationReason: field$1.deprecationReason,
      description: field$1.description,
      extensions: field$1.extensions,

      resolve(rootVal) {
        for (var _len = arguments.length, extra = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          extra[_key - 1] = arguments[_key];
        }

        if (rootVal.value === null) {
          return innerResolver(rootVal, ...extra);
        }

        let value = null;

        try {
          value = JSON.parse(rootVal.value);
        } catch (err) {}

        return innerResolver({
          item: rootVal.item,
          value
        }, ...extra);
      }

    })
  }).value;
}

function mapUpdateInputArgToSQLite(arg) {
  if (arg === undefined) {
    return undefined;
  }

  return {
    arg: arg.arg,

    async resolve(input, context, relationshipInputResolver) {
      const resolvedInput = arg.resolve === undefined ? input : await arg.resolve(input, context, relationshipInputResolver);

      if (resolvedInput === undefined || resolvedInput === null) {
        return resolvedInput;
      }

      return JSON.stringify(resolvedInput);
    }

  };
}

function mapCreateInputArgToSQLite(arg) {
  if (arg === undefined) {
    return undefined;
  }

  return {
    arg: arg.arg,

    async resolve(input, context, relationshipInputResolver) {
      const resolvedInput = arg.resolve === undefined ? input : await arg.resolve(input, context, relationshipInputResolver);

      if (resolvedInput === undefined || resolvedInput === null) {
        return resolvedInput;
      }

      return JSON.stringify(resolvedInput);
    }

  };
}

function jsonFieldTypePolyfilledForSQLite(provider, config, dbFieldConfig) {
  var _dbFieldConfig$mode2;

  if (provider === 'sqlite') {
    var _dbFieldConfig$mode, _config$input, _config$input2;

    return fieldType({
      kind: 'scalar',
      mode: (_dbFieldConfig$mode = dbFieldConfig === null || dbFieldConfig === void 0 ? void 0 : dbFieldConfig.mode) !== null && _dbFieldConfig$mode !== void 0 ? _dbFieldConfig$mode : 'optional',
      scalar: 'String',
      default: dbFieldConfig === null || dbFieldConfig === void 0 ? void 0 : dbFieldConfig.default,
      map: dbFieldConfig === null || dbFieldConfig === void 0 ? void 0 : dbFieldConfig.map
    })(_objectSpread(_objectSpread({}, config), {}, {
      input: {
        create: mapCreateInputArgToSQLite((_config$input = config.input) === null || _config$input === void 0 ? void 0 : _config$input.create),
        update: mapUpdateInputArgToSQLite((_config$input2 = config.input) === null || _config$input2 === void 0 ? void 0 : _config$input2.update)
      },
      output: mapOutputFieldToSQLite(config.output),
      extraOutputFields: Object.fromEntries(Object.entries(config.extraOutputFields || {}).map(_ref2 => {
        let [key, field] = _ref2;
        return [key, mapOutputFieldToSQLite(field)];
      }))
    }));
  }

  return fieldType({
    kind: 'scalar',
    mode: (_dbFieldConfig$mode2 = dbFieldConfig === null || dbFieldConfig === void 0 ? void 0 : dbFieldConfig.mode) !== null && _dbFieldConfig$mode2 !== void 0 ? _dbFieldConfig$mode2 : 'optional',
    scalar: 'Json',
    default: dbFieldConfig === null || dbFieldConfig === void 0 ? void 0 : dbFieldConfig.default,
    map: dbFieldConfig === null || dbFieldConfig === void 0 ? void 0 : dbFieldConfig.map
  })(config);
}

export { resolveString as a, filters as f, jsonFieldTypePolyfilledForSQLite as j, resolveCommon as r };
