// @keystone-6/core/dist/mysql-673b47d5.esm.js
import { D as DateTime$3, a as Decimal$3, B as BigInt$3 } from './graphql-ts-schema-57db0426.esm.js';
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import 'decimal.js';
import '@graphql-tools/schema';
import { enum as enum$1, enumValues, inputObject, arg, list, nonNull, String as String$3, Boolean as Boolean$3, Int as Int$3, Float as Float$3 } from '@graphql-ts/schema/api-without-context';

const orderDirectionEnum = enum$1({
  name: 'OrderDirection',
  values: enumValues(['asc', 'desc'])
});
const QueryMode = enum$1({
  name: 'QueryMode',
  values: enumValues(['default', 'insensitive'])
});
// fieldType(dbField)(fieldInfo) => { ...fieldInfo, dbField };
function fieldType(dbField) {
  return function (graphQLInfo) {
    return _objectSpread(_objectSpread({}, graphQLInfo), {}, {
      dbField
    });
  };
}

// (even though, yes, having EnumFilter by defined as EnumNullableFilter<Enum>, would be the same type but names would show up differently in editors for example)

function enumFilters(enumType) {
  const optional = inputObject({
    name: `${enumType.graphQLType.name}NullableFilter`,
    fields: () => ({
      equals: arg({
        type: enumType
      }),
      in: arg({
        type: list(nonNull(enumType))
      }),
      notIn: arg({
        type: list(nonNull(enumType))
      }),
      not: arg({
        type: optional
      })
    })
  });
  const required = inputObject({
    name: `${enumType.graphQLType.name}Filter`,
    fields: () => ({
      equals: arg({
        type: enumType
      }),
      in: arg({
        type: list(nonNull(enumType))
      }),
      notIn: arg({
        type: list(nonNull(enumType))
      }),
      not: arg({
        type: optional
      })
    })
  });
  const many = inputObject({
    name: `${enumType.graphQLType.name}NullableListFilter`,
    fields: () => ({
      // can be null
      equals: arg({
        type: list(nonNull(enumType))
      }),
      // can be null
      has: arg({
        type: enumType
      }),
      hasEvery: arg({
        type: list(nonNull(enumType))
      }),
      hasSome: arg({
        type: list(nonNull(enumType))
      }),
      isEmpty: arg({
        type: enumType
      })
    })
  });
  return {
    optional,
    required,
    many
  };
}

// Do not manually modify this file, it is automatically generated by the package at /prisma-utils in this repo.
const StringNullableFilter$2 = inputObject({
  name: 'StringNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: String$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(String$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    mode: arg({
      type: QueryMode
    }),
    // can be null
    not: arg({
      type: NestedStringNullableFilter$2
    })
  })
});
const NestedStringNullableFilter$2 = inputObject({
  name: 'NestedStringNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: String$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(String$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    // can be null
    not: arg({
      type: NestedStringNullableFilter$2
    })
  })
});
const StringFilter$2 = inputObject({
  name: 'StringFilter',
  fields: () => ({
    equals: arg({
      type: String$3
    }),
    in: arg({
      type: list(nonNull(String$3))
    }),
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    mode: arg({
      type: QueryMode
    }),
    not: arg({
      type: NestedStringFilter$2
    })
  })
});
const NestedStringFilter$2 = inputObject({
  name: 'NestedStringFilter',
  fields: () => ({
    equals: arg({
      type: String$3
    }),
    in: arg({
      type: list(nonNull(String$3))
    }),
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    not: arg({
      type: NestedStringFilter$2
    })
  })
});
const StringNullableListFilter = inputObject({
  name: 'StringNullableListFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: list(nonNull(String$3))
    }),
    // can be null
    has: arg({
      type: String$3
    }),
    hasEvery: arg({
      type: list(nonNull(String$3))
    }),
    hasSome: arg({
      type: list(nonNull(String$3))
    }),
    isEmpty: arg({
      type: Boolean$3
    })
  })
});
const BoolNullableFilter$2 = inputObject({
  name: 'BooleanNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Boolean$3
    }),
    // can be null
    not: arg({
      type: BoolNullableFilter$2
    })
  })
});
const BoolFilter$2 = inputObject({
  name: 'BooleanFilter',
  fields: () => ({
    equals: arg({
      type: Boolean$3
    }),
    not: arg({
      type: BoolFilter$2
    })
  })
});
const BoolNullableListFilter = inputObject({
  name: 'BooleanNullableListFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: list(nonNull(Boolean$3))
    }),
    // can be null
    has: arg({
      type: Boolean$3
    }),
    hasEvery: arg({
      type: list(nonNull(Boolean$3))
    }),
    hasSome: arg({
      type: list(nonNull(Boolean$3))
    }),
    isEmpty: arg({
      type: Boolean$3
    })
  })
});
const IntNullableFilter$2 = inputObject({
  name: 'IntNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Int$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Int$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Int$3))
    }),
    lt: arg({
      type: Int$3
    }),
    lte: arg({
      type: Int$3
    }),
    gt: arg({
      type: Int$3
    }),
    gte: arg({
      type: Int$3
    }),
    // can be null
    not: arg({
      type: IntNullableFilter$2
    })
  })
});
const IntFilter$2 = inputObject({
  name: 'IntFilter',
  fields: () => ({
    equals: arg({
      type: Int$3
    }),
    in: arg({
      type: list(nonNull(Int$3))
    }),
    notIn: arg({
      type: list(nonNull(Int$3))
    }),
    lt: arg({
      type: Int$3
    }),
    lte: arg({
      type: Int$3
    }),
    gt: arg({
      type: Int$3
    }),
    gte: arg({
      type: Int$3
    }),
    not: arg({
      type: IntFilter$2
    })
  })
});
const IntNullableListFilter = inputObject({
  name: 'IntNullableListFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: list(nonNull(Int$3))
    }),
    // can be null
    has: arg({
      type: Int$3
    }),
    hasEvery: arg({
      type: list(nonNull(Int$3))
    }),
    hasSome: arg({
      type: list(nonNull(Int$3))
    }),
    isEmpty: arg({
      type: Boolean$3
    })
  })
});
const FloatNullableFilter$2 = inputObject({
  name: 'FloatNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Float$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Float$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Float$3))
    }),
    lt: arg({
      type: Float$3
    }),
    lte: arg({
      type: Float$3
    }),
    gt: arg({
      type: Float$3
    }),
    gte: arg({
      type: Float$3
    }),
    // can be null
    not: arg({
      type: FloatNullableFilter$2
    })
  })
});
const FloatFilter$2 = inputObject({
  name: 'FloatFilter',
  fields: () => ({
    equals: arg({
      type: Float$3
    }),
    in: arg({
      type: list(nonNull(Float$3))
    }),
    notIn: arg({
      type: list(nonNull(Float$3))
    }),
    lt: arg({
      type: Float$3
    }),
    lte: arg({
      type: Float$3
    }),
    gt: arg({
      type: Float$3
    }),
    gte: arg({
      type: Float$3
    }),
    not: arg({
      type: FloatFilter$2
    })
  })
});
const FloatNullableListFilter = inputObject({
  name: 'FloatNullableListFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: list(nonNull(Float$3))
    }),
    // can be null
    has: arg({
      type: Float$3
    }),
    hasEvery: arg({
      type: list(nonNull(Float$3))
    }),
    hasSome: arg({
      type: list(nonNull(Float$3))
    }),
    isEmpty: arg({
      type: Boolean$3
    })
  })
});
const DateTimeNullableFilter$2 = inputObject({
  name: 'DateTimeNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: DateTime$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(DateTime$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(DateTime$3))
    }),
    lt: arg({
      type: DateTime$3
    }),
    lte: arg({
      type: DateTime$3
    }),
    gt: arg({
      type: DateTime$3
    }),
    gte: arg({
      type: DateTime$3
    }),
    // can be null
    not: arg({
      type: DateTimeNullableFilter$2
    })
  })
});
const DateTimeFilter$2 = inputObject({
  name: 'DateTimeFilter',
  fields: () => ({
    equals: arg({
      type: DateTime$3
    }),
    in: arg({
      type: list(nonNull(DateTime$3))
    }),
    notIn: arg({
      type: list(nonNull(DateTime$3))
    }),
    lt: arg({
      type: DateTime$3
    }),
    lte: arg({
      type: DateTime$3
    }),
    gt: arg({
      type: DateTime$3
    }),
    gte: arg({
      type: DateTime$3
    }),
    not: arg({
      type: DateTimeFilter$2
    })
  })
});
const DateTimeNullableListFilter = inputObject({
  name: 'DateTimeNullableListFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: list(nonNull(DateTime$3))
    }),
    // can be null
    has: arg({
      type: DateTime$3
    }),
    hasEvery: arg({
      type: list(nonNull(DateTime$3))
    }),
    hasSome: arg({
      type: list(nonNull(DateTime$3))
    }),
    isEmpty: arg({
      type: Boolean$3
    })
  })
});
const DecimalNullableFilter$2 = inputObject({
  name: 'DecimalNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Decimal$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Decimal$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Decimal$3))
    }),
    lt: arg({
      type: Decimal$3
    }),
    lte: arg({
      type: Decimal$3
    }),
    gt: arg({
      type: Decimal$3
    }),
    gte: arg({
      type: Decimal$3
    }),
    // can be null
    not: arg({
      type: DecimalNullableFilter$2
    })
  })
});
const DecimalFilter$2 = inputObject({
  name: 'DecimalFilter',
  fields: () => ({
    equals: arg({
      type: Decimal$3
    }),
    in: arg({
      type: list(nonNull(Decimal$3))
    }),
    notIn: arg({
      type: list(nonNull(Decimal$3))
    }),
    lt: arg({
      type: Decimal$3
    }),
    lte: arg({
      type: Decimal$3
    }),
    gt: arg({
      type: Decimal$3
    }),
    gte: arg({
      type: Decimal$3
    }),
    not: arg({
      type: DecimalFilter$2
    })
  })
});
const DecimalNullableListFilter = inputObject({
  name: 'DecimalNullableListFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: list(nonNull(Decimal$3))
    }),
    // can be null
    has: arg({
      type: Decimal$3
    }),
    hasEvery: arg({
      type: list(nonNull(Decimal$3))
    }),
    hasSome: arg({
      type: list(nonNull(Decimal$3))
    }),
    isEmpty: arg({
      type: Boolean$3
    })
  })
});
const BigIntNullableFilter$2 = inputObject({
  name: 'BigIntNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: BigInt$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(BigInt$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(BigInt$3))
    }),
    lt: arg({
      type: BigInt$3
    }),
    lte: arg({
      type: BigInt$3
    }),
    gt: arg({
      type: BigInt$3
    }),
    gte: arg({
      type: BigInt$3
    }),
    // can be null
    not: arg({
      type: BigIntNullableFilter$2
    })
  })
});
const BigIntFilter$2 = inputObject({
  name: 'BigIntFilter',
  fields: () => ({
    equals: arg({
      type: BigInt$3
    }),
    in: arg({
      type: list(nonNull(BigInt$3))
    }),
    notIn: arg({
      type: list(nonNull(BigInt$3))
    }),
    lt: arg({
      type: BigInt$3
    }),
    lte: arg({
      type: BigInt$3
    }),
    gt: arg({
      type: BigInt$3
    }),
    gte: arg({
      type: BigInt$3
    }),
    not: arg({
      type: BigIntFilter$2
    })
  })
});
const BigIntNullableListFilter = inputObject({
  name: 'BigIntNullableListFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: list(nonNull(BigInt$3))
    }),
    // can be null
    has: arg({
      type: BigInt$3
    }),
    hasEvery: arg({
      type: list(nonNull(BigInt$3))
    }),
    hasSome: arg({
      type: list(nonNull(BigInt$3))
    }),
    isEmpty: arg({
      type: Boolean$3
    })
  })
});
const String$2 = {
  optional: StringNullableFilter$2,
  required: StringFilter$2,
  many: StringNullableListFilter
};
const Boolean$2 = {
  optional: BoolNullableFilter$2,
  required: BoolFilter$2,
  many: BoolNullableListFilter
};
const Int$2 = {
  optional: IntNullableFilter$2,
  required: IntFilter$2,
  many: IntNullableListFilter
};
const Float$2 = {
  optional: FloatNullableFilter$2,
  required: FloatFilter$2,
  many: FloatNullableListFilter
};
const DateTime$2 = {
  optional: DateTimeNullableFilter$2,
  required: DateTimeFilter$2,
  many: DateTimeNullableListFilter
};
const Decimal$2 = {
  optional: DecimalNullableFilter$2,
  required: DecimalFilter$2,
  many: DecimalNullableListFilter
};
const BigInt$2 = {
  optional: BigIntNullableFilter$2,
  required: BigIntFilter$2,
  many: BigIntNullableListFilter
};

var postgresql = /*#__PURE__*/Object.freeze({
  __proto__: null,
  String: String$2,
  Boolean: Boolean$2,
  Int: Int$2,
  Float: Float$2,
  DateTime: DateTime$2,
  Decimal: Decimal$2,
  BigInt: BigInt$2,
  'enum': enumFilters
});

// Do not manually modify this file, it is automatically generated by the package at /prisma-utils in this repo.
const StringNullableFilter$1 = inputObject({
  name: 'StringNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: String$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(String$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    // can be null
    not: arg({
      type: NestedStringNullableFilter$1
    })
  })
});
const NestedStringNullableFilter$1 = inputObject({
  name: 'NestedStringNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: String$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(String$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    // can be null
    not: arg({
      type: NestedStringNullableFilter$1
    })
  })
});
const StringFilter$1 = inputObject({
  name: 'StringFilter',
  fields: () => ({
    equals: arg({
      type: String$3
    }),
    in: arg({
      type: list(nonNull(String$3))
    }),
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    not: arg({
      type: NestedStringFilter$1
    })
  })
});
const NestedStringFilter$1 = inputObject({
  name: 'NestedStringFilter',
  fields: () => ({
    equals: arg({
      type: String$3
    }),
    in: arg({
      type: list(nonNull(String$3))
    }),
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    not: arg({
      type: NestedStringFilter$1
    })
  })
});
const BoolNullableFilter$1 = inputObject({
  name: 'BooleanNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Boolean$3
    }),
    // can be null
    not: arg({
      type: BoolNullableFilter$1
    })
  })
});
const BoolFilter$1 = inputObject({
  name: 'BooleanFilter',
  fields: () => ({
    equals: arg({
      type: Boolean$3
    }),
    not: arg({
      type: BoolFilter$1
    })
  })
});
const IntNullableFilter$1 = inputObject({
  name: 'IntNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Int$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Int$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Int$3))
    }),
    lt: arg({
      type: Int$3
    }),
    lte: arg({
      type: Int$3
    }),
    gt: arg({
      type: Int$3
    }),
    gte: arg({
      type: Int$3
    }),
    // can be null
    not: arg({
      type: IntNullableFilter$1
    })
  })
});
const IntFilter$1 = inputObject({
  name: 'IntFilter',
  fields: () => ({
    equals: arg({
      type: Int$3
    }),
    in: arg({
      type: list(nonNull(Int$3))
    }),
    notIn: arg({
      type: list(nonNull(Int$3))
    }),
    lt: arg({
      type: Int$3
    }),
    lte: arg({
      type: Int$3
    }),
    gt: arg({
      type: Int$3
    }),
    gte: arg({
      type: Int$3
    }),
    not: arg({
      type: IntFilter$1
    })
  })
});
const FloatNullableFilter$1 = inputObject({
  name: 'FloatNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Float$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Float$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Float$3))
    }),
    lt: arg({
      type: Float$3
    }),
    lte: arg({
      type: Float$3
    }),
    gt: arg({
      type: Float$3
    }),
    gte: arg({
      type: Float$3
    }),
    // can be null
    not: arg({
      type: FloatNullableFilter$1
    })
  })
});
const FloatFilter$1 = inputObject({
  name: 'FloatFilter',
  fields: () => ({
    equals: arg({
      type: Float$3
    }),
    in: arg({
      type: list(nonNull(Float$3))
    }),
    notIn: arg({
      type: list(nonNull(Float$3))
    }),
    lt: arg({
      type: Float$3
    }),
    lte: arg({
      type: Float$3
    }),
    gt: arg({
      type: Float$3
    }),
    gte: arg({
      type: Float$3
    }),
    not: arg({
      type: FloatFilter$1
    })
  })
});
const DateTimeNullableFilter$1 = inputObject({
  name: 'DateTimeNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: DateTime$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(DateTime$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(DateTime$3))
    }),
    lt: arg({
      type: DateTime$3
    }),
    lte: arg({
      type: DateTime$3
    }),
    gt: arg({
      type: DateTime$3
    }),
    gte: arg({
      type: DateTime$3
    }),
    // can be null
    not: arg({
      type: DateTimeNullableFilter$1
    })
  })
});
const DateTimeFilter$1 = inputObject({
  name: 'DateTimeFilter',
  fields: () => ({
    equals: arg({
      type: DateTime$3
    }),
    in: arg({
      type: list(nonNull(DateTime$3))
    }),
    notIn: arg({
      type: list(nonNull(DateTime$3))
    }),
    lt: arg({
      type: DateTime$3
    }),
    lte: arg({
      type: DateTime$3
    }),
    gt: arg({
      type: DateTime$3
    }),
    gte: arg({
      type: DateTime$3
    }),
    not: arg({
      type: DateTimeFilter$1
    })
  })
});
const DecimalNullableFilter$1 = inputObject({
  name: 'DecimalNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Decimal$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Decimal$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Decimal$3))
    }),
    lt: arg({
      type: Decimal$3
    }),
    lte: arg({
      type: Decimal$3
    }),
    gt: arg({
      type: Decimal$3
    }),
    gte: arg({
      type: Decimal$3
    }),
    // can be null
    not: arg({
      type: DecimalNullableFilter$1
    })
  })
});
const DecimalFilter$1 = inputObject({
  name: 'DecimalFilter',
  fields: () => ({
    equals: arg({
      type: Decimal$3
    }),
    in: arg({
      type: list(nonNull(Decimal$3))
    }),
    notIn: arg({
      type: list(nonNull(Decimal$3))
    }),
    lt: arg({
      type: Decimal$3
    }),
    lte: arg({
      type: Decimal$3
    }),
    gt: arg({
      type: Decimal$3
    }),
    gte: arg({
      type: Decimal$3
    }),
    not: arg({
      type: DecimalFilter$1
    })
  })
});
const BigIntNullableFilter$1 = inputObject({
  name: 'BigIntNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: BigInt$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(BigInt$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(BigInt$3))
    }),
    lt: arg({
      type: BigInt$3
    }),
    lte: arg({
      type: BigInt$3
    }),
    gt: arg({
      type: BigInt$3
    }),
    gte: arg({
      type: BigInt$3
    }),
    // can be null
    not: arg({
      type: BigIntNullableFilter$1
    })
  })
});
const BigIntFilter$1 = inputObject({
  name: 'BigIntFilter',
  fields: () => ({
    equals: arg({
      type: BigInt$3
    }),
    in: arg({
      type: list(nonNull(BigInt$3))
    }),
    notIn: arg({
      type: list(nonNull(BigInt$3))
    }),
    lt: arg({
      type: BigInt$3
    }),
    lte: arg({
      type: BigInt$3
    }),
    gt: arg({
      type: BigInt$3
    }),
    gte: arg({
      type: BigInt$3
    }),
    not: arg({
      type: BigIntFilter$1
    })
  })
});
const String$1 = {
  optional: StringNullableFilter$1,
  required: StringFilter$1
};
const Boolean$1 = {
  optional: BoolNullableFilter$1,
  required: BoolFilter$1
};
const Int$1 = {
  optional: IntNullableFilter$1,
  required: IntFilter$1
};
const Float$1 = {
  optional: FloatNullableFilter$1,
  required: FloatFilter$1
};
const DateTime$1 = {
  optional: DateTimeNullableFilter$1,
  required: DateTimeFilter$1
};
const Decimal$1 = {
  optional: DecimalNullableFilter$1,
  required: DecimalFilter$1
};
const BigInt$1 = {
  optional: BigIntNullableFilter$1,
  required: BigIntFilter$1
};

var sqlite = /*#__PURE__*/Object.freeze({
  __proto__: null,
  String: String$1,
  Boolean: Boolean$1,
  Int: Int$1,
  Float: Float$1,
  DateTime: DateTime$1,
  Decimal: Decimal$1,
  BigInt: BigInt$1,
  'enum': enumFilters
});

// Do not manually modify this file, it is automatically generated by the package at /prisma-utils in this repo.
const StringNullableFilter = inputObject({
  name: 'StringNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: String$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(String$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    // can be null
    not: arg({
      type: NestedStringNullableFilter
    })
  })
});
const NestedStringNullableFilter = inputObject({
  name: 'NestedStringNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: String$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(String$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    // can be null
    not: arg({
      type: NestedStringNullableFilter
    })
  })
});
const StringFilter = inputObject({
  name: 'StringFilter',
  fields: () => ({
    equals: arg({
      type: String$3
    }),
    in: arg({
      type: list(nonNull(String$3))
    }),
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    not: arg({
      type: NestedStringFilter
    })
  })
});
const NestedStringFilter = inputObject({
  name: 'NestedStringFilter',
  fields: () => ({
    equals: arg({
      type: String$3
    }),
    in: arg({
      type: list(nonNull(String$3))
    }),
    notIn: arg({
      type: list(nonNull(String$3))
    }),
    lt: arg({
      type: String$3
    }),
    lte: arg({
      type: String$3
    }),
    gt: arg({
      type: String$3
    }),
    gte: arg({
      type: String$3
    }),
    contains: arg({
      type: String$3
    }),
    startsWith: arg({
      type: String$3
    }),
    endsWith: arg({
      type: String$3
    }),
    not: arg({
      type: NestedStringFilter
    })
  })
});
const BoolNullableFilter = inputObject({
  name: 'BooleanNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Boolean$3
    }),
    // can be null
    not: arg({
      type: BoolNullableFilter
    })
  })
});
const BoolFilter = inputObject({
  name: 'BooleanFilter',
  fields: () => ({
    equals: arg({
      type: Boolean$3
    }),
    not: arg({
      type: BoolFilter
    })
  })
});
const IntNullableFilter = inputObject({
  name: 'IntNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Int$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Int$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Int$3))
    }),
    lt: arg({
      type: Int$3
    }),
    lte: arg({
      type: Int$3
    }),
    gt: arg({
      type: Int$3
    }),
    gte: arg({
      type: Int$3
    }),
    // can be null
    not: arg({
      type: IntNullableFilter
    })
  })
});
const IntFilter = inputObject({
  name: 'IntFilter',
  fields: () => ({
    equals: arg({
      type: Int$3
    }),
    in: arg({
      type: list(nonNull(Int$3))
    }),
    notIn: arg({
      type: list(nonNull(Int$3))
    }),
    lt: arg({
      type: Int$3
    }),
    lte: arg({
      type: Int$3
    }),
    gt: arg({
      type: Int$3
    }),
    gte: arg({
      type: Int$3
    }),
    not: arg({
      type: IntFilter
    })
  })
});
const FloatNullableFilter = inputObject({
  name: 'FloatNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Float$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Float$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Float$3))
    }),
    lt: arg({
      type: Float$3
    }),
    lte: arg({
      type: Float$3
    }),
    gt: arg({
      type: Float$3
    }),
    gte: arg({
      type: Float$3
    }),
    // can be null
    not: arg({
      type: FloatNullableFilter
    })
  })
});
const FloatFilter = inputObject({
  name: 'FloatFilter',
  fields: () => ({
    equals: arg({
      type: Float$3
    }),
    in: arg({
      type: list(nonNull(Float$3))
    }),
    notIn: arg({
      type: list(nonNull(Float$3))
    }),
    lt: arg({
      type: Float$3
    }),
    lte: arg({
      type: Float$3
    }),
    gt: arg({
      type: Float$3
    }),
    gte: arg({
      type: Float$3
    }),
    not: arg({
      type: FloatFilter
    })
  })
});
const DateTimeNullableFilter = inputObject({
  name: 'DateTimeNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: DateTime$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(DateTime$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(DateTime$3))
    }),
    lt: arg({
      type: DateTime$3
    }),
    lte: arg({
      type: DateTime$3
    }),
    gt: arg({
      type: DateTime$3
    }),
    gte: arg({
      type: DateTime$3
    }),
    // can be null
    not: arg({
      type: DateTimeNullableFilter
    })
  })
});
const DateTimeFilter = inputObject({
  name: 'DateTimeFilter',
  fields: () => ({
    equals: arg({
      type: DateTime$3
    }),
    in: arg({
      type: list(nonNull(DateTime$3))
    }),
    notIn: arg({
      type: list(nonNull(DateTime$3))
    }),
    lt: arg({
      type: DateTime$3
    }),
    lte: arg({
      type: DateTime$3
    }),
    gt: arg({
      type: DateTime$3
    }),
    gte: arg({
      type: DateTime$3
    }),
    not: arg({
      type: DateTimeFilter
    })
  })
});
const DecimalNullableFilter = inputObject({
  name: 'DecimalNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: Decimal$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(Decimal$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(Decimal$3))
    }),
    lt: arg({
      type: Decimal$3
    }),
    lte: arg({
      type: Decimal$3
    }),
    gt: arg({
      type: Decimal$3
    }),
    gte: arg({
      type: Decimal$3
    }),
    // can be null
    not: arg({
      type: DecimalNullableFilter
    })
  })
});
const DecimalFilter = inputObject({
  name: 'DecimalFilter',
  fields: () => ({
    equals: arg({
      type: Decimal$3
    }),
    in: arg({
      type: list(nonNull(Decimal$3))
    }),
    notIn: arg({
      type: list(nonNull(Decimal$3))
    }),
    lt: arg({
      type: Decimal$3
    }),
    lte: arg({
      type: Decimal$3
    }),
    gt: arg({
      type: Decimal$3
    }),
    gte: arg({
      type: Decimal$3
    }),
    not: arg({
      type: DecimalFilter
    })
  })
});
const BigIntNullableFilter = inputObject({
  name: 'BigIntNullableFilter',
  fields: () => ({
    // can be null
    equals: arg({
      type: BigInt$3
    }),
    // can be null
    in: arg({
      type: list(nonNull(BigInt$3))
    }),
    // can be null
    notIn: arg({
      type: list(nonNull(BigInt$3))
    }),
    lt: arg({
      type: BigInt$3
    }),
    lte: arg({
      type: BigInt$3
    }),
    gt: arg({
      type: BigInt$3
    }),
    gte: arg({
      type: BigInt$3
    }),
    // can be null
    not: arg({
      type: BigIntNullableFilter
    })
  })
});
const BigIntFilter = inputObject({
  name: 'BigIntFilter',
  fields: () => ({
    equals: arg({
      type: BigInt$3
    }),
    in: arg({
      type: list(nonNull(BigInt$3))
    }),
    notIn: arg({
      type: list(nonNull(BigInt$3))
    }),
    lt: arg({
      type: BigInt$3
    }),
    lte: arg({
      type: BigInt$3
    }),
    gt: arg({
      type: BigInt$3
    }),
    gte: arg({
      type: BigInt$3
    }),
    not: arg({
      type: BigIntFilter
    })
  })
});
const String = {
  optional: StringNullableFilter,
  required: StringFilter
};
const Boolean = {
  optional: BoolNullableFilter,
  required: BoolFilter
};
const Int = {
  optional: IntNullableFilter,
  required: IntFilter
};
const Float = {
  optional: FloatNullableFilter,
  required: FloatFilter
};
const DateTime = {
  optional: DateTimeNullableFilter,
  required: DateTimeFilter
};
const Decimal = {
  optional: DecimalNullableFilter,
  required: DecimalFilter
};
const BigInt = {
  optional: BigIntNullableFilter,
  required: BigIntFilter
};

var mysql = /*#__PURE__*/Object.freeze({
  __proto__: null,
  String: String,
  Boolean: Boolean,
  Int: Int,
  Float: Float,
  DateTime: DateTime,
  Decimal: Decimal,
  BigInt: BigInt,
  'enum': enumFilters
});

export { QueryMode as Q, fieldType as f, mysql as m, orderDirectionEnum as o, postgresql as p, sqlite as s };
