// @keystone-6/core/dist/useCreateItem-676a5df7.esm.js
import { useToasts } from '@keystone-ui/toast';
import { useState, useMemo, useRef, useEffect, useCallback } from 'react';
import isDeepEqual from 'fast-deep-equal';
import { useMutation, gql } from '@apollo/client';
import { useKeystone } from '../admin-ui/context/dist/keystone-6-core-admin-ui-context.esm.js';
import { u as usePreventNavigation } from './GraphQLErrorNotice-28a3062f.esm.js';

function useCreateItem(list) {
  const toasts = useToasts();
  const {
    createViewFieldModes
  } = useKeystone();
  const [createItem, {
    loading,
    error,
    data: returnedData
  }] = useMutation(gql`mutation($data: ${list.gqlNames.createInputName}!) {
      item: ${list.gqlNames.createMutationName}(data: $data) {
        id
        label: ${list.labelField}
    }
  }`);
  const [value, setValue] = useState(() => {
    const value = {};
    Object.keys(list.fields).forEach(fieldPath => {
      value[fieldPath] = {
        kind: 'value',
        value: list.fields[fieldPath].controller.defaultValue
      };
    });
    return value;
  });
  const invalidFields = useMemo(() => {
    const invalidFields = new Set();
    Object.keys(value).forEach(fieldPath => {
      const val = value[fieldPath].value;
      const validateFn = list.fields[fieldPath].controller.validate;

      if (validateFn) {
        const result = validateFn(val);

        if (result === false) {
          invalidFields.add(fieldPath);
        }
      }
    });
    return invalidFields;
  }, [list, value]);
  const [forceValidation, setForceValidation] = useState(false);
  const data = {};
  Object.keys(list.fields).forEach(fieldPath => {
    const {
      controller
    } = list.fields[fieldPath];
    const serialized = controller.serialize(value[fieldPath].value);

    if (!isDeepEqual(serialized, controller.serialize(controller.defaultValue))) {
      Object.assign(data, serialized);
    }
  });
  const shouldPreventNavigation = !(returnedData !== null && returnedData !== void 0 && returnedData.item) && Object.keys(data).length !== 0;
  const shouldPreventNavigationRef = useRef(shouldPreventNavigation);
  useEffect(() => {
    shouldPreventNavigationRef.current = shouldPreventNavigation;
  }, [shouldPreventNavigation]);
  usePreventNavigation(shouldPreventNavigationRef);
  return {
    state: loading ? 'loading' : !(returnedData !== null && returnedData !== void 0 && returnedData.item) ? 'created' : 'editing',
    shouldPreventNavigation,
    error,
    props: {
      fields: list.fields,
      fieldModes: createViewFieldModes.state === 'loaded' ? createViewFieldModes.lists[list.key] : null,
      forceValidation,
      invalidFields,
      value,
      onChange: useCallback(getNewValue => {
        setValue(oldValues => getNewValue(oldValues));
      }, [])
    },

    async create() {
      const newForceValidation = invalidFields.size !== 0;
      setForceValidation(newForceValidation);
      if (newForceValidation) return undefined;
      let outputData;

      try {
        outputData = await createItem({
          variables: {
            data
          }
        }).then(x => x.data);
      } catch {
        return undefined;
      }

      shouldPreventNavigationRef.current = false;
      const label = outputData.item.label || outputData.item.id;
      toasts.addToast({
        title: label,
        message: 'Created Successfully',
        tone: 'positive'
      });
      return outputData.item;
    }

  };
}

export { useCreateItem as u };
