// @contentlayer/source-files/dist/fetchData/DocumentContext.js
import * as tracing_1 from "@effect-ts/core/Tracing";
const fileName_1 = "packages/@contentlayer/source-files/src/fetchData/DocumentContext.ts";
import * as utils from '@contentlayer/utils';
import { T, tag } from '@contentlayer/utils/effect';
import * as path from 'node:path';
import { getFlattenedPath } from './mapping.js';
export const DocumentContext = tag(Symbol.for('@contentlayer/source-files/DocumentContext'));
export const provideDocumentContext = (_) => T.provideService(DocumentContext)(_, fileName_1 + ":21:96");
export const makeAndProvideDocumentContext = ({ rawContent, relativeFilePath, }) => {
    const contentType = utils.pattern
        .match(rawContent.kind)
        .with('markdown', () => 'markdown')
        .with('mdx', () => 'mdx')
        .otherwise(() => 'data');
    const rawDocumentData = {
        sourceFilePath: relativeFilePath,
        sourceFileName: path.basename(relativeFilePath),
        sourceFileDir: path.dirname(relativeFilePath),
        contentType,
        flattenedPath: getFlattenedPath(relativeFilePath),
    };
    return provideDocumentContext({ rawContent, rawDocumentData, relativeFilePath });
};
export const getFromDocumentContext = (key) => T.accessService(DocumentContext)((_) => _[key], fileName_1 + ":44:35");
