// @keystone-6/core/___internal-do-not-use-will-break-in-patch/next-graphql/dist/keystone-6-core-___internal-do-not-use-will-break-in-patch-next-graphql.esm.js
import { i as initConfig, c as createSystem } from '../../../dist/initConfig-5d2c2305.esm.js';
import { c as createApolloServerMicro } from '../../../dist/createApolloServer-23a6674d.esm.js';
import '@babel/runtime/helpers/objectSpread2';
import 'p-limit';
import '../../../dist/core-c6bc4160.esm.js';
import '../../../dist/mysql-673b47d5.esm.js';
import '../../../dist/graphql-ts-schema-57db0426.esm.js';
import '@graphql-ts/schema';
import 'graphql-type-json';
import 'graphql-upload/GraphQLUpload.js';
import 'graphql';
import 'decimal.js';
import '@graphql-ts/schema/api-without-context';
import '@graphql-ts/extend';
import '@graphql-ts/schema/api-with-context';
import '@graphql-tools/schema';
import '@babel/runtime/helpers/objectWithoutProperties';
import '../../../dist/graphql-errors-68845105.esm.js';
import 'apollo-server-errors';
import 'path';
import 'util';
import 'fs-extra';
import 'resolve';
import '@nodelib/fs.walk';
import '@emotion/hash';
import '../../../dist/admin-meta-graphql-9585711c.esm.js';
import '@apollo/client';
import '@babel/runtime/helpers/classPrivateFieldInitSpec';
import '@babel/runtime/helpers/classPrivateFieldGet';
import '@babel/runtime/helpers/classPrivateFieldSet';
import '../../../dist/types-for-lists-9414dddb.esm.js';
import 'graphql/execution/values';
import 'pluralize';
import '@babel/runtime/helpers/defineProperty';
import 'uuid';
import 'image-type';
import 'image-size';
import 'stream';
import '@aws-sdk/s3-request-presigner';
import '@aws-sdk/client-s3';
import '@aws-sdk/lib-storage';
import 'crypto';
import 'filenamify';
import '@sindresorhus/slugify';
import 'cuid';
import '../../../dist/package-path-5ec371a3.esm.js';
import 'apollo-server-micro';
import 'apollo-server-express';
import 'apollo-server-core';
import '../../../session/dist/keystone-6-core-session.esm.js';
import 'cookie';
import '@hapi/iron';
import 'uid-safe';

function nextGraphQLAPIRoute(keystoneConfig, prismaClient) {
  const initializedKeystoneConfig = initConfig(keystoneConfig);
  const {
    graphQLSchema,
    getKeystone
  } = createSystem(initializedKeystoneConfig);
  const keystone = getKeystone(prismaClient);
  keystone.connect();
  const apolloServer = createApolloServerMicro({
    graphQLSchema,
    createContext: keystone.createContext,
    sessionStrategy: initializedKeystoneConfig.session,
    graphqlConfig: initializedKeystoneConfig.graphql,
    connectionPromise: keystone.connect()
  });
  let startPromise = apolloServer.start();
  return async (req, res) => {
    var _keystoneConfig$graph;

    await startPromise;
    return apolloServer.createHandler({
      path: ((_keystoneConfig$graph = keystoneConfig.graphql) === null || _keystoneConfig$graph === void 0 ? void 0 : _keystoneConfig$graph.path) || '/api/graphql'
    })(req, res);
  };
}

export { nextGraphQLAPIRoute };
