// @keystone-ui/options/dist/keystone-ui-options.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import _objectWithoutProperties from '@babel/runtime/helpers/esm/objectWithoutProperties';
import _extends from '@babel/runtime/helpers/esm/extends';
import { jsx, useTheme } from '@keystone-ui/core';
import { useIndicatorTokens } from '@keystone-ui/fields';
import { CheckIcon } from '@keystone-ui/icons/icons/CheckIcon';
import { useMemo } from 'react';
import ReactSelect, { components } from 'react-select';

const _excluded = ["selectProps"],
      _excluded2 = ["components"];
const CheckMark = _ref => {
  let {
    isDisabled,
    isFocused,
    isSelected
  } = _ref;
  const tokens = useIndicatorTokens({
    size: 'medium',
    type: 'radio'
  });
  return jsx("div", {
    className: `${isDisabled ? 'disabled ' : ''}${isFocused ? 'focus ' : ''}${isSelected ? 'selected' : ''}`,
    css: {
      alignItems: 'center',
      backgroundColor: tokens.background,
      borderColor: tokens.borderColor,
      borderRadius: tokens.borderRadius,
      borderStyle: 'solid',
      borderWidth: tokens.borderWidth,
      boxSizing: 'border-box',
      color: tokens.foreground,
      cursor: 'pointer',
      display: 'flex',
      flexShrink: 0,
      height: tokens.boxSize,
      justifyContent: 'center',
      transition: tokens.transition,
      width: tokens.boxSize,
      '&.focus': {
        backgroundColor: tokens.focus.background,
        borderColor: tokens.focus.borderColor,
        boxShadow: tokens.focus.shadow,
        color: tokens.focus.foreground
      },
      '&.selected': {
        backgroundColor: tokens.selected.background,
        borderColor: tokens.selected.borderColor,
        boxShadow: tokens.selected.shadow,
        color: tokens.selected.foreground
      },
      '&.disabled': {
        backgroundColor: tokens.disabled.background,
        borderColor: tokens.disabled.borderColor,
        boxShadow: tokens.disabled.shadow,
        color: tokens.disabled.background,
        cursor: 'default'
      },
      '&.selected.disabled': {
        color: tokens.disabled.foreground
      }
    }
  }, jsx(CheckIcon, {
    size: "small"
  }));
};
const OptionPrimitive = _ref2 => {
  let {
    children,
    isDisabled,
    isFocused,
    innerProps,
    innerRef,
    className
  } = _ref2;
  const theme = useTheme();
  return jsx("div", _extends({
    ref: innerRef,
    className: className,
    css: {
      alignItems: 'center',
      color: isDisabled ? theme.colors.foregroundDim : isFocused ? theme.colors.linkHoverColor : undefined,
      cursor: 'pointer',
      display: 'flex',
      fontSize: '0.9em',
      fontWeight: 500,
      justifyContent: 'space-between',
      background: isFocused ? theme.colors.backgroundHover : undefined,
      outline: 0,
      padding: `${theme.spacing.small}px`,
      pointerEvents: isDisabled ? 'none' : undefined,
      '&:not(:first-of-type)': {
        borderTop: `1px solid ${theme.colors.backgroundDim}`
      },
      ':hover': {
        background: theme.colors.backgroundHover,
        color: theme.colors.linkHoverColor
      }
    }
  }, innerProps), children);
};

const Control = _ref3 => {
  let {
    selectProps
  } = _ref3,
      props = _objectWithoutProperties(_ref3, _excluded);

  return jsx(components.Control, _extends({
    selectProps: selectProps
  }, props));
};

const defaultComponents = {
  Control,
  Option: OptionPrimitive,
  DropdownIndicator: null,
  IndicatorSeparator: null
};
const Options = _ref4 => {
  let {
    components: propComponents
  } = _ref4,
      props = _objectWithoutProperties(_ref4, _excluded2);

  const components = useMemo(() => _objectSpread(_objectSpread({}, defaultComponents), propComponents), [propComponents]);
  const theme = useTheme();
  const optionRendererStyles = useMemo(() => ({
    control: provided => _objectSpread(_objectSpread({}, provided), {}, {
      background: theme.fields.inputBackground,
      boxShadow: 'none',
      cursor: 'text',
      padding: 0,
      minHeight: 34
    }),
    menu: () => ({
      marginTop: 8
    }),
    menuList: provided => _objectSpread(_objectSpread({}, provided), {}, {
      padding: 0
    }),
    placeholder: provided => _objectSpread(_objectSpread({}, provided), {}, {
      color: theme.fields.inputPlaceholder
    })
  }), [theme]);
  return jsx(ReactSelect, _extends({
    backspaceRemovesValue: false,
    captureMenuScroll: false,
    closeMenuOnSelect: false,
    controlShouldRenderValue: false,
    hideSelectedOptions: false,
    isClearable: false,
    isSearchable: true,
    maxMenuHeight: 1000,
    menuIsOpen: true,
    menuShouldScrollIntoView: false,
    styles: optionRendererStyles // TODO: JW: Not a fan of this, but it doesn't seem to make a difference
    // if we take it out. react-select bug maybe?
    ,
    tabSelectsValue: false,
    components: components
  }, props));
};

export { CheckMark, OptionPrimitive, Options };
